### DISCUSSION TITLE
Can't export privatekeys using generate_keys

### DISCUSSION THREAD

**Discussion by nishaanth-ms-21155** [sparkle-project/Sparkle#2624] [Q&A] [answered]
Using Sparkle 1.24 to and am using generate_keys -x my_private_key and it still saves it in keychain. But it works in Sparkle's latest distribution provided. Is the only way to upgrade Sparkle in my app? If so, how tiring will the API changes

**Comment by zorgiepoo** [comment, marked as answer]
You can use the latest tools even if you're using an older framework in your app. Check https://github.com/sparkle-project/Sparkle/discussions/2248

For upgrading the Sparkle framework (recommended) check https://sparkle-project.org/documentation/upgrading/

There shouldn't be any breaking API changes from 1.24 -> 1.27.3 so you should be able to easily upgrade to that version at minimum.

Upgrading to Sparkle 2 is a bit more work but still worthwhile (e.g. more efficient updating). There's a compatibility API layer however so you may be able to upgrade without making changes, initially (albeit with deprecation warnings). Check the upgrading page above for more detailed information.

**Reply by nishaanth-ms-21155** [reply]
Thanks for your help. Can I ask you one more question? I remember seeing some discussions outside of GitHub where delta updates can apply one after the other when there exists a delta for every adjacent versions. Is that true? If so from which version that is supported?
Thanks in advance.

**Reply by zorgiepoo** [reply]
Look under [Manual generation](https://sparkle-project.org/documentation/delta-updates#manual-generation) of delta updates documentation to see how it works in detail. You can provide multiple delta update files for a new release, and only one is selected; which one is selected depends on which version of your app the user is updating from.

Sparkle's delta updates have always worked this way. Multiple delta updates are not applied one after the other, and it's not needed.

**Reply by nishaanth-ms-21155** [reply]
Yes, I have applied the delta through the app and it successfully updates. But had questions multiple delta updates for a single update through chaining existing consecutive deltas.

**Reply by zorgiepoo** [reply]
Read my above reply. You are misunderstanding how Sparkle's delta updates work (there's no daisy chaining) and I've already explained the approach Sparkle uses instead above for delta updating from different older versions.
